test_plans/reference_semantics.asciidoc

```diff
@@ -0,0 +1,148 @@
+- `host_unsampled_image_accessor<int>`
+- `kernel`
+- `kernel_id`
+- `kernel_bundle<bundle_state::executable>`
+- `platform`
+- `queue`
+- `sampled_image`
+- `unsampled_image`
+
+The following "kernel types" are identified:
+
+- `accessor<int, 1, access_mode::read_write, target::device>`
+- `local_accessor<int, 0>`
+- `unsampled_image_accessor<int, 1, access_mode::read>`
+- `sampled_image_accessor<int, 1>`
+- `stream`
+
+=== Host application tests
+
+These tests are run from a host application for all host types and kernel types.
+
+==== Equality via copy, symmetry
+
+This test is repeated for the copy constructor and for copy assignment. For class `T`:
+
+1. Create an instance `t0` of `T`.
+2. Create a copy of `t0`, `t1`.
+3. Check that `t0 == t1` and `t1 == t0`.
+4. Check that `std::hash<T>{}(t0) == std::hash<T>{}(t1)`.
+
+==== Non-equality via copy, symmetry
+
+This test is not executed for: `device_image`, `device` if only one device is available and `platform` if only one platform is available. For class `T`:
+
+1. Create an instance `t0` of `T`.
+2. Create an instance `t1` of `T`, which is distinct from `t0`.
+3. Check that `t0 != t1` and `t1 != t0`.
+4. Check that `std::hash<T>{}(t0) != std::hash<T>{}(t1)`.
+
+==== Reflexivity
+
+For class `T`:
+
+1. Create an instance `t0` of `T`.
+2. Check that `t0 == t0`.
+3. Check that `std::hash<T>{}(t0) == std::hash<T>{}(t0)`.
+
+==== Transitivity
+
+This test is repeated for the copy constructor and for copy assignment. For class `T`:
+
+1. Create an instance `t0` of `T`.
+2. Create a copy of `t0`, `t1`.
+3. Create a copy of `t1`, `t2`.
+4. Check that `t0 == t2` and `t2 == t0`.
+5. Check that `std::hash<T>{}(t0) == std::hash<T>{}(t2)`.
+
+=== Shared tests
+
+These tests are run both from a host application for all host types and kernel types, and from a kernel function for all kernel types. Some of the below tests require mutation of an instance `T`. This is not always possible. When it is not possible, do not execute those tests.
+
+==== Type traits
+
+For class `T`, check that the following statements evaluate to `true`.
+
+- `std::is_copy_constructible_v<T>`.
+- `std::is_copy_assignable_v<T>`
+- `std::is_destructible_v<T>`
+- `std::is_move_constructible_v<T>`
+- `std::is_move_assignable_v<T>`
+
+==== Copy
+
+This test is repeated for the copy constructor and for copy assignment. For class `T`:
+
+1. Create an instance `t0` of `T`.
+2. Store the properties of `t0` by calling its member functions.
+3. Create a copy of `t0`, `t1`.
+4. Check that the member functions of `t1` return the same values as stored in Step 2.
+
+==== Const-correctness
+Check that a `const` copy of an instance is not affected by the mutations to the instance that is copied.
+
+1. Create an instance `t0` of `T`.
+2. Make a `const` copy of `t0`, `t1`.
+3. Store the properties of `t1` by calling its member functions.
+4. Mutate `t0`.
+5. Check that the member functions of `t1` return the same values as stored in Step 2.
```

Comment:
```suggestion
5. Check that the member functions of `t1` return the same values as stored in Step 3.
```

I assume "step 2" was just a typo.

This test looks wrong to me, though.  [Section 4.5.2](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:reference-semantics) makes this guarantee about objects with common reference semantics:

> Any instance of T that is constructed as a copy of another instance, via either the copy constructor or copy assignment operator, **must behave as-if it were the original instance and as-if any action performed on it were also performed on the original instance** ...

There is no exception here for `const` objects.  In addition, the spec guarantees that `t0 == t1`, so it would be very weird if `t0` and `t1` didn't return the same values from their member functions.

Reply:
Yes, indeed a typo. Okay I see, I have changed this test to check that a "mutation" is also applied to the `const` copy.

Interestingly, the test in `common_by_reference.h` is:
```c++
  const auto &const_obj = a;
  storage.store(const_obj);
  mutator(a, mutation::const_correctness); // changes instance a
  // failure if const_obj not the same as what was stored before the mutation
  util::run_check(log, mkInfo("Const-correcntess."), [&]{
    if (!storage.check(const_obj)) { 
      FAIL(log, testName + " ignores const correctness (storage::check())");
    }
  });
```

Unless I am missing something, this test is wrong.

Reply:
Yes, that does seem wrong.  I wonder if that test failed before?

In any case, the new test plan you have written looks correct w.r.t. `const` copies.

Reply:
No, the `common_by_reference.h` header is only used by `kernel_id`, which passes a no-op `mutator`.

I must add that it is (apparently) hard to get the tests for the reference semantics right with regards to the propagation of changes. However, I feel that this is quite a crucial feature of these semantics.